packages/sitecore-jss-nextjs/src/editing/editing-render-middleware.ts

```diff
@@ -241,44 +224,211 @@ export class EditingRenderMiddleware extends RenderMiddlewareBase {
+    } as EditingMetadataPreviewData);
+
+    const route = this.config.resolvePageUrl?.({ itemPath: query.route }) || query.route;
+
+    debug.editing(
+      'editing render middleware end in %dms: redirect %o',
+      Date.now() - startTimestamp,
+      {
+        status: 307,
+        route,
+      }
     );
+
+    // Restrict the page to be rendered only within the allowed origins
+    res.setHeader('Content-Security-Policy', this.getSCPHeader());
+
+    res.redirect(route);
+  }
+
+  /**
+   * Gets the Content-Security-Policy header value
+   * @returns Content-Security-Policy header value
+   */
+  getSCPHeader() {
+    return `frame-ancestors 'self' ${[getAllowedOriginsFromEnv(), ...EDITING_ALLOWED_ORIGINS].join(
+      ' '
+    )}`;
+  }
+}
+
+/**
+ * Middleware / handler for use in the editing render Next.js API route (e.g. '/api/editing/render')
+ * which is required for Sitecore editing support.
+ */
+export class EditingRenderMiddleware extends RenderMiddlewareBase {
+  /**
+   * @param {EditingRenderMiddlewareConfig} [config] Editing render middleware config
+   */
+  constructor(public config?: EditingRenderMiddlewareConfig) {
+    super();
   }
 
-  const layoutData = JSON.parse(payload.args[1]);
-  const viewBag = JSON.parse(payload.args[2]);
-  // Keep backwards compatibility in case people use an older JSS version that doesn't send the path in the context
-  const path = layoutData.sitecore.context.itemPath ?? viewBag.httpContext.request.path;
+  /**
+   * Gets the Next.js API route handler
+   * @returns route handler
+   */
+  public getHandler(): (req: NextApiRequest, res: NextApiResponse) => Promise<void> {
+    return this.handler;
+  }
+
+  private handler = async (req: NextApiRequest, res: NextApiResponse): Promise<void> => {
+    const { query, body, method, headers } = req;
 
-  return {
-    path,
-    layoutData,
-    language: viewBag.language,
-    dictionary: viewBag.dictionary,
+    debug.editing('editing render middleware start: %o', {
+      method,
+      query,
+      headers,
+      body,
+    });
+
+    if (!enforceCors(req, res, EDITING_ALLOWED_ORIGINS)) {
+      debug.editing(
+        'invalid origin host - set allowed origins in JSS_ALLOWED_ORIGINS environment variable'
+      );
+      return res.status(401).json({
+        html: `<html><body>Requests from origin ${req.headers?.origin} not allowed</body></html>`,
+      });
+    }
+
+    // Validate secret
+    const secret = query[QUERY_PARAM_EDITING_SECRET] ?? body?.jssEditingSecret;
+    if (secret !== getJssEditingSecret()) {
+      debug.editing(
+        'invalid editing secret - sent "%s" expected "%s"',
+        secret,
+        getJssEditingSecret()
+      );
+      return res.status(401).json({
+        html: '<html><body>Missing or invalid secret</body></html>',
+      });
+    }
+
+    switch (req.method) {
+      case 'GET': {
+        const handler = new MetadataHandler({ resolvePageUrl: this.config?.resolvePageUrl });
```

Comment:
Do we want to add any validation of the query params (i.e. return 400 if any essentials are missing)?

Reply:
Makes sense. Added